proposals/0033-timely-vote-credits.md

```diff
@@ -0,0 +1,389 @@
+(voted_in_slot - voted_on_slot).
+
+These vote latencies are stored in u8 field of the Lockout data structure
+named *latency*.
+
+The credits awarded for a rooted slot are calculated using the latency value
+stored in the latency field the slot, using a formula that awards latencies
+of 1 - 2 slots 8 credits, with a 1 credit reduction for each vote latency
+after 2.  Rooted slots will always be awarded a minimum credit of 1 (never
+0) so that very old votes, possibly necessary in times of network stress,
+are not discouraged.
+
+The implementation of the above is fairly simple for Vote transactions,
+because Vote transactions include slot numbers voted on, and that list of
+slots is already pruned of already-voted-on slots; so each newly voted-on
+slot that survives this pruning can have its latency calculated simply.
+
+VoteStateUpdate is more complex to implement since it is not merely a list
+of slots to vote on.  When a VoteStateUpdate instruction is processed:
+
+1. For each slot that was in the previous VoteState but are not in the new
+VoteState because they have been rooted in the transition from the old
+VoteState to the new VoteState, credits to award are calculated based on the
+latency that was recorded for them and still available in the old VoteState.
+
+2. For each slot that was in both the previous VoteState and the new
+VoteState, the latency that was previously recorded for that slot is copied
+from the old VoteState to the new VoteState.
+
+3. For each slot that is in the new VoteState but wasn't in the old
+VoteState, the latency value is calculated for this new slot according to
+what slot the vote is for and what slot is in the Clock (i.e. the slot this
+VoteStateUpdate tx landed in) and this latency is stored in VoteState for
+that slot.
+
+The code to handle this is more complex, because VoteStateUpdate may include
+removal of slots that expired as performed by the voting validator, in
+addition to slots that have been rooted and new slots added.  However, the
+assumptions that are needed to handle VoteStateUpdate with timely vote
+credits are already guaranteed by existing VoteStateUpdate correctness
+checking code:
+
+The existing VoteStateUpdate processing code already ensures that (1) only
+roots slots that could actually have been rooted in the transition from the
+old VoteState to the new VoteState, so there is no danger of over-counting
+credits (i.e. imagine that a 'cheating' validator "pretended" that slots
+were rooted by dropping them off of the back of the new VoteState before
+they have actually achieved 32 confirmations; the existing logic prevents
+this).
+
+The existing VoteStateUpdate processing code already ensures that (2) new
+slots included in the new VoteState are only slots after slots that have
+already been voted on in the old VoteState (i.e. can't inject new slots in
+the middle of slots already voted on).
+
+## Impact
+
+1. Any code which deserialized vote account state will require a recompile
```

Comment:
I think instead for this bullet point you should outline the feature flag for upgrading the new vote state format (and implications on any rent balance) as well as the non impact on the tower.

Reply:
Actually I see you talked about it down below, maybe just a note here that the on chain vote state format will change and see the `Backwards Compatibility` section for the upgrade procedure.